### PR TITLE
Gandi DNS service

### PR DESCRIPTION
**Pull request by alanbacon**
Thank you for contributing to sewer.                    
Every contribution to sewer is important to us.                       
You may not know it, but you have just contributed to making the world a more safer and secure place.                         

Contributor offers to license certain software (a “Contribution” or multiple
“Contributions”) to sewer, and sewer agrees to accept said Contributions,
under the terms of the MIT License.
Contributor understands and agrees that sewer shall have the irrevocable and perpetual right to make
and distribute copies of any Contribution, as well as to create and distribute collective works and
derivative works of any Contribution, under the MIT License.

Now,                   

## What(What have you changed?)

Added a dns_provider class for the gandi DNS service
+ tests (although the requests mock is a bit hacky, see comments)
+ readme

## Why(Why did you change it?)

Because I use gandi DNS and it isn't currently supported.

### CODE REVIEW

File: sewer/dns_providers/tests/test_gandi.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            ])
+
+        return MockResponseObject(status_code=404)
+
+    mock_requests_get.side_effect = mock_implementation
+
+
+def add_side_effect_to_request_post(mock_requests_post):
+    def mock_implementation(*args, **kwargs):
+        assert kwargs['headers'] == {
```
Comment (komuw) [inline sewer/dns_providers/tests/test_gandi.py:46]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5350192)

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- sewer/dns_providers/gandi.py  3
- sewer/dns_providers/tests/test_gandi.py  4
- sewer/cli.py  2
         

Clones added
============
- sewer/dns_providers/tests/test_gandi.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5350192&bid=17520538)

File: sewer/dns_providers/tests/test_gandi.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            return MockResponseObject()
+
+        return MockResponseObject(status_code=404)
+
+    mock_requests_post.side_effect = mock_implementation
+
+
+def add_side_effect_to_request_delete(mock_requests_delete):
+    def mock_implementation(*args, **kwargs):
+        assert kwargs['headers'] == {"X-Api-Key": MOCK_GANDI_API_KEY}, 'Get headers not set'
```
Comment (komuw) [inline sewer/dns_providers/tests/test_gandi.py:61]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5350192)

File: sewer/dns_providers/gandi.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    """
+
+    dns_provider_name = "gandi"
+
+    def __init__(
+        self,
+        GANDI_API_KEY=None,
+        GANDI_API_BASE_URL="https://dns.api.gandi.net/api/v5/",
+        DEFAULT_TTL=10800,
+        requests_lib=requests
```
Comment (alanbacon) [inline sewer/dns_providers/gandi.py:18]: There is an option to pass in a requests library here so that I can pass in a mock requests library. I do this because I was struggling to get python to apply my mock correctly in this file (yes I have googled it a lot). If you have any suggetions I'll happily listen.

File: sewer/dns_providers/tests/test_gandi.py (hunk 1 above)
Comment (alanbacon) [reply sewer/dns_providers/tests/test_gandi.py:46]: thanks, ill change that bit

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- sewer/cli.py  2
- sewer/dns_providers/tests/test_gandi.py  3
- sewer/dns_providers/gandi.py  3
         

Clones added
============
- sewer/dns_providers/tests/test_gandi.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5350192&bid=17520538)

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- sewer/cli.py  2
- sewer/dns_providers/gandi.py  3
- sewer/dns_providers/tests/test_gandi.py  3
         

Clones added
============
- sewer/dns_providers/tests/test_gandi.py  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5350192&bid=17520538)

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- sewer/dns_providers/tests/test_gandi.py  3
- sewer/dns_providers/gandi.py  3
- sewer/cli.py  2
         

Clones added
============
- sewer/dns_providers/tests/test_gandi.py  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5350192&bid=17520538)

**Comment by alanbacon** [comment]
when i run black locally over the files I have added it says that the files will be left unchanged, however circleci says the files will be reformatted.... presumably resulting in a build failure. Can you help me with this? Thanks

**Comment by mmaney** [comment]
@alanbacon are you running black with `-l 100`?  It's obsessive line breaking/wrapping is quite unstable to changes in target line lengths.

**Comment by alanbacon** [comment]
@mmaney yes I ran it with the same arguments that I saw in the failed circleci check:
```
black --line-length=100 --check
```
but with no luck

is there a pipfile.lock or anything that I can install a specfic version from?

**Comment by mmaney** [comment]
Are you sure black is looking at the test_gandi.py file?  I can see some lines over 100 chars in there that it ought to be reformatting.

**Comment by mmaney** [comment]
@alanbacon Well, that's a nuisance... if there's an easier way to pull your gandi branch than making a separate checkout...

Anyway, with that done:

`.../tests$ black -l 100 --check test_gandi.py `
`would reformat test_gandi.py`
`All done! 💥 💔 💥`
`1 file would be reformatted.`

black, version 18.9b0 - which is stock for Debian stable, so not especially new & improved.

Most of the changes it makes - and IME this version agrees with the CI version - are removing trailing commas from the last argument to function calls.  I have a vague recollection that this was something that can be changed by an option, since that didn't used to be valid Python.  Oh yeah, --py36.  Which seems to only be able to be turned on by an option, you have to turn it off by removing the commas so it won't auto-detect that you want that style, which is stupid.  Stupid, stupid, stupid.  It's so stupid that maybe you have a later version where they've corrected that oversight?

(I am not a fan of black, and even if I had been, having to de-format the with-stacks in the existing tests so they could be read by a human would have cured me of it.  It's as bad as lint (the famous Bug: there are things that you cannot make lint shut up about) but worse because with black you can't make it leave code readable.  I say it's spinach...)

Here's the as-is (in 224c) of those stupid comma lines:

-            get_calls[0][0][0], "https://dns.api.gandi.net/api/v5/domains/second-level-domain.tld",
-            get_calls[0][0][0], "https://dns.api.gandi.net/api/v5/domains/second-level-domain.tld",
-            get_calls[2][0][0], "https://dns.api.gandi.net/api/v5/domains/second-level-domain.tld",
-            post_calls[0][1]["json"]["rrset_name"], "_acme-challenge.subsubdomain.subdomain",
-            get_calls[0][0][0], "https://dns.api.gandi.net/api/v5/domains/second-level-domain.tld",
-            get_calls[2][0][0], "https://dns.api.gandi.net/api/v5/domains/second-level-domain.tld",

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- sewer/dns_providers/gandi.py  3
- sewer/dns_providers/tests/test_gandi.py  3
- sewer/cli.py  2
         

Clones added
============
- sewer/dns_providers/tests/test_gandi.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5350192&bid=17520538)

**Comment by alanbacon** [comment]
@mmaney thanks, i formatted with python3.6 and black====18.9b0 and everything is passing now. Thanks

**Comment by alanbacon** [comment]
@mmaney @komuw  so can this be merged now?

**Comment by mmaney** [comment]
@alanbacon @komuw I am of two minds about it.  I need to preface the exposition of this schizophrenic condition by saying that despite my wordy assistance with the formatting issues, I hadn't really examined the code until a couple days ago.

Mind #1: Alan says it works for him, code is no worse IMO than the already accepted dns providers, so I guess we could commit it as-is.  I'm pretty sure that's as much QC as most of the providers have had... and we've patched the problems as they get reported.  #168, #162, #134, #132, ...

Mind #2: there are a lot of things I would like to see cleaned up, but let me say it again: yes, this would be holding Alan's code to a higher standard than a lot of what's already in the tree.  Two examples (I have a little list...):

2.1: gandi would have to be added to the list in #164, since it uses what looks a lot like the incorrect (or we could say "of limited scope") code in a couple other providers.  Also note that I haven't come up with an entirely satisfactory solution for this.  :-(

2.2: There's a lot of unnecessary, even counterproductive fragmentation of the code into trivial methods and nested functions.  And staticmethods, yuck.

Some of this is a matter of taste, but... consider subdomain_to_challenge_domain(subdomain), which is entirely pointless IMO.  staticmethod, check.  The computation it does is so trivial that it's probably a little shorter coded inline, and it hides nothing that's worth hiding.

This sort of thing could be cleaned up later, but I would much prefer to see it done while the author, who can test that it actually works with gandi, is still involved and motivated to get it shined up before it's committed.  :-)

**Comment by alanbacon** [comment]
@mmaney @komuw 

I think at the root of both the staticmethods problems is that the API to the DNS providers isn't quite right:

Instead of:
```
def create_dns_record(self, domain_name, domain_dns_value):
  pass
```

It would be better if it was:

```
def create_dns_record(self, challenge_domain, base_domain, domain_dns_value):
  pass
```

This would make the DNS_provider classes much simpler. There wouldn't have to be any string manipulation (e.g. splitting subdomains) in the DNS_provider classes at all.

This would have the effect of getting the user to specify a dict of the sub_domains they want to get certificates for, a feature I would quite like.

```
{
 "sub_domains": [<sub_domains>],
 "base_domain": <base_domain>
}
```

Also with a changed API: each DNS_provider class woudn't have to append `"_acme-challenge"` to the domain, which at the moment they do, meaning that there is repeated code where there doesn't need to be. Instead it would be the responsibility of the the code that calls the provider to append the `"_acme-challenge"` part. This also has another benefit, if letsEncrypt change their API at some point from `_acme-challenge` to (for example) `_le_new_challenge_path` then the DNS_providers classes will still dumb enough to be reused.

---

TL;DR 

I agree static methods are not nice, it indicates the fact that both those methods do not belong in that class. But they do belong somewhere, and that somewhere is probably the main logic in `client.py`.

If the DNS_provider classes were simpler: then we (as code maintainers*) wouldn't be as responsible for maintaining large classes that were submitted as PRs. Classes that are difficult for us to test, because we probably only actually use one of them.

[* assuming you will endulge me that title for the time being]

Flip side: that refactor is probably too large to undertake now (because it will affect every DNS_provider class). And it certaintly doesn't belong in this PR.

**Comment by alanbacon** [comment]
### Defending small functions:

As for the small methods, as you said yourself it's a matter of taste. So to defend my taste: at work we have a coding standard where small functions are good because they make the code "self-documenting" and easier to read. Also small functions are easier to unit test as there are less paths to confuse things.

Take for example this code:

```
def has_second_level_domain(split_domains):
    return len(split_domains) >= 2

if not has_second_level_domain(split_domains):
            ...
```

I personally feel that is easier to read and understand what is happening rather than the shorter but equivelent code:

```
if not len(split_domains) >= 2:
            ...
```

Which now gives you zero context about what is happening

**Comment by mmaney** [comment]
@alanbacon you're not the first to think about adding the explicit zone apex to the dns provider interface!  But as you recognize, that would be too invasive to force on all the existing ones.

If you don't mind waiting a little longer, I'm currently working on a refactoring that would make just such additional parameters, whether shared or provider-specific, very easy to handle.  #161 is now somewhat out of date, but the core idea for the new interface is in there.  x_cert have become just x, and x_auth has been dropped as not needed.  I would *love* to have gandi be the first new-style dns provider in the tree!

(that version of the all-provider API was abandonded as conflicting changes and changes in my understanding of what was needed both made it seem easier to recreate it than to pile patches on top)

**Comment by alanbacon** [comment]
In the spirit of being agile I would prefer to get this merged and closed off, then once your #161 changes are made I will happily open a new PR to update the gandi dns provider class to use your new interface.

At that point we could start thinking about how to deprecate the old styles in a backwards compitble way. I'm happy to help with that and let the gandi dns provider class be the guinea pig for testing out how that transition will work.

By leaving PRs dangling, I find they just become stale and difficult to keep up-to-date, plus it's all to easy to end up in a situation where nothing is ever merged because every PR ends up waiting on another PR.

**Comment by mmaney** [comment]
Okay, we'll go with what's there for now.

BTW, I've remembered one problem with assuming the user can provide a base_domain: there's no requirement that all the domains in a certificate HAVE a single common base.  No idea how common (or unheard of) this might be.  :-/

**Comment by alanbacon** [comment]
sweet, thanks 👍

**Comment by alanbacon** [comment]
Oh I've never come across that problem because I have a separate cert for each of my sub domains and domains, didn't actually know there was an alternative to this.

**Comment by mmaney** [comment]
Yeah, I think I came across a mention of actually having "unrelated" domains in one cert, but it was a while ago.  That reminds me, in gandi you seem to assume there can be only one record for any fqdn.  And in many cases that's correct, but there's a special case where it is definitely not true.  If one wants a single certificate to cover *.domain.tld and the naked domain.tld (which is often made to be both a service endpoint as well as a DNS domain), then ACME will need to verify both for the certificate, and challenges must be attached to domain.tld.  That does complicate things for some DNS service providers, and sewer has the bugs to show that people really do want to do this.  :-/

Life would be so much easier if everyone just ran bind locally as the domain master and used services only as slaves... from the viewpoint of anyone maintaining an ACME client like sewer, anwyay.  :-)

**Comment by mmaney** [comment]
@alanbacon ah, missed answering this bit before - it's illustrative of things that would be nice if only...

`Also with a changed API: each DNS_provider class woudn't have to append "_acme-challenge" to the domain, which at the moment they do, meaning that there is repeated code where there doesn't need to be. Instead it would be the responsibility of the code that calls the provider (client.py) to append the "_acme-challenge" part`

You mean "prepends" there `<wink>`.  And that would be a good idea except for one thing: there are times when that's NOT how you form the name of the DNS record.  It is a less common option, but say you have a public DNS service that you don't really wish to leave, but they don't have an API for the ACME client to frob (or perhaps their system isn't designed for fast propagation and you don't want to wait five minutes to be sure the server will see the challenge data), then you might want to use a sneaky CNAME approach:

_acme-challenge.host.domain.tld is setup statically as a CNAME to host.acme_validation.domain.tld, where acme_validation.domain.tld is served by a more readily controlled DNS server where your ACME client can publish the challenges easily and quickly.  Obviously, handing the client _acme-challenge.host.domain.tld is of no benefit in this case.  (this was the use case where I too thought "if I could just tell it what the apex domain is...".  Cheated and hardwired it into the unpublished code to get things working.)

Also, http-01 challenges, which sewer just gained support for (no real working providers yet, but doon I'm sure), don't use the _acme-challenges prefix at all, so this would push provider-type switching back into the protocol code... nope, don't want to do that.  This sort of thing, along with the base_domain splitting (which isn't needed at all by some dns providers!) might be candidates for a shared implementation in lib, or the dns common.py.

Which further reminds me that you were asking about the path forward?  I get the auth/providers refactoring in, with compatibility layers for legacy dns providers (hopefully there won't yet be any legacy http providers) and they get ported to the new API when and as necessary.  Or, possibly, dropped, if there's no one who can dig into its service-specific problems around.

Oh, and get the RFC8555 protocol changes landed, too.  November gets closer every day.

**Comment by alanbacon** [comment]
> You mean "prepends" there <wink>. 

"Prepend" yes sorry, my grammer was poor.

---

Surely in the case you have highlighted where two providers are needed; rather than using a CNAME you would work it like this:

```
Provider 1 (slow / no api):      host.domain.tld  10800 A <some_static_ip>

Provider 2 (fast / has api) :    _acme-challenge.host.domain.tld 10800 TXT <challenge text>
```

Which simplifies things back to the standard use case again.

---

> If one wants a single certificate to cover *.domain.tld and the naked domain.tld (which is often made to be both a service endpoint as well as a DNS domain), then ACME will need to verify both for the certificate, and challenges must be attached to domain.tld

Could you explain this a bit more, what is the process for getting a certificate that will work with '*.domain.tld"?

And what additional steps need to be taken if one also wants that cert to cover naked "domain.tld"?

Is there a single challenge for each? That then results in a single cert?

If so that should still just be two calls to  `create_dns_record`.

---

> Also, http-01 challenges, which sewer just gained support for (no real working providers yet, but soon I'm sure), don't use the _acme-challenges prefix at all, so this would push provider-type switching back into the protocol code... nope, don't want to do that. This sort of thing, along with the base_domain splitting (which isn't needed at all by some dns providers!) might be candidates for a shared implementation in lib, or the dns common.py.

So my personal view point throughout the following is that DNS-provider classes should be as thin and reuable as possible, so that as easy as possible to maintain by either the maintainers of the this project, or by the contributers that original submitted them.

Essentially:
```
create_dns_record(challenge_subdomain, apex_domain, txt_value_to_write)
```

1. I understand that splitting the domains is not needed by all the providers, but sticking known strings together is generally easier than spliting them apart and trying to work out the semantics.

e.g at the moment you can pass in the cli args:

```
--domain example.tld --alt_domains host.example.tld
```

This means the some code somewhere has to work the out the semantics of what the apex domain is and what the challenge domain is.

Wouldn't it be easier to write the code if the semantics were more clearly passed in?:

```
--apex example.tld --challenges "" "host" "*"
```

or written as JSON:

```
{
  "apex": "example.tld",
  "challenges": [
    ""
    "host",
    "*"
  ]
}
```

2. Http challanges are currently run as a separate class, is your PR thinking of changing that?

3. Why wouldn't you want to have provider type switching logic in `client.py`? Where should it live, not in the dns-provider class surely?

**Comment by mmaney** [comment]
`Surely in the case you have highlighted where two providers are needed; rather than using a CNAME you would work it like this:`

No, the CNAME trick is required because if provider1 is authoritative for d.t, the server will never look at provider2.  Nor will anyone else unless they point their resolver at provider2.  That's how DNS works.

`If so that should still just be two calls to create_dns_record.`

And that is exactly what happens.  The problem is that in these calls, different challenge values need to be added as TXT records for the same _acme-challenge.d.t, because "*" isn't a thing in DNS (invalid character to be in a name part).  And as we've seen, apparently whenever someone tries to create such a certificate, the services' API doesn't do what we'd like.  I haven't had to try to fix one up yet, but from what I couldn't avoid reading, some need a different API function for adding the second TXT record, and at least one sounded like it needed the user to submit a multi-valued create... challenging if the driver gets the challenges one at a time.

Clearing the challenges can be another can of worms.  Again, the legacy API makes two calls for the same name, but the service API seems about equally likely to delete all the TXT values on the first call and error on the second.  Or something else that the driver has to work around in a service-specific way.

`Wouldn't it be easier to write the code if the semantics were more clearly passed in?:`

Sure.  But that's not the user interface we have now, and frankly I hope sewer never tries to force its users to do use such an unnatural (for normal humans, not developers) interface just to save us writing a few lines of code.  :-/

`Why wouldn't you want to have provider type switching logic in client.py? Where should it live, not in the dns-provider class surely?`

Because type-switching code is inherently fragile and easily broken.  (and wasn't that one of the motivations behind OOP in the first place? <wink>)  The switching is done by choosing the provider(1); they are then tasked with hiding the differences.  Some differences are common to a whole category (dns-01 uses a hash of the key_auth; http-01 uses it directly).  Now, if someone wanted to survey all the existing dns providers and determine d that nearly all of them used a splitting like the buggy "domain.tld, it's always just two parts, right?" or the more accurate tldextract (or maybe something else), then that could be reason to hoist that operation into the base class.  Eventually.  For the moment, there are quite a few legacy providers that expect to be given the fqdn in one string, and we can't just abandon them on short notice.

In the long run, sure, my own inclination would be to demote to "contrib" any driver that lacks an active developer who can actually test it against the service provider.  Slightly shady, but still, probably, working.

(1) so have you looked at the incredible mess of provider-aware code in cli.py?  Doesn't that make you want to vomit?  It was one of the very first things I noticed when I started digging into the sewer code back in January (before that I only dabbled lightly), and I have a Wonderful Plan to get rid of it... but it's still working, more or less, and so on my list it's behind getting sewer working with the lastest protocol changes and other things.
